File: server/hotrod/src/test/scala/org/infinispan/server/hotrod/HotRodAuthenticationTest.scala
package org.infinispan.server.hotrod

import java.lang.reflect.Method
import java.util.HashMap
import javax.security.sasl.Sasl

import io.netty.channel.group.ChannelGroup
import org.infinispan.manager.EmbeddedCacheManager
import org.infinispan.server.core.security.simple.SimpleServerAuthenticationProvider
import org.infinispan.server.hotrod.configuration.HotRodServerConfigurationBuilder
import org.infinispan.server.hotrod.test.HotRodTestingUtil.startHotRodServer
import org.infinispan.server.hotrod.test.{HotRodTestingUtils, TestCallbackHandler}
import org.infinispan.test.TestingUtil
import org.jboss.sasl.JBossSaslProvider
import org.testng.AssertJUnit.{assertEquals, assertTrue}
import org.testng.annotations.Test

/**
 * Hot Rod server authentication test.
 *
 * @author Tristan Tarrant
 * @since 7.0
 */
@Test(groups = Array("functional"), testName = "server.hotrod.HotRodAuthenticationTest")
class HotRodAuthenticationTest extends HotRodSingleNodeTest {
   val jbossSaslProvider = new JBossSaslProvider()

   override def createStartHotRodServer(cacheManager: EmbeddedCacheManager): HotRodServer = {
      val ssap = new SimpleServerAuthenticationProvider
      ssap.addUser("user", "realm", "password".toCharArray, null)
      val builder = new HotRodServerConfigurationBuilder
      builder.authentication().enable().addAllowedMech("CRAM-MD5").serverAuthenticationProvider(ssap).serverName("localhost").addMechProperty(Sasl.POLICY_NOANONYMOUS, "true")
      startHotRodServer(cacheManager, HotRodTestingUtils.serverPort, 0, builder)
   }

   def testAuthMechList(m: Method) {
      val a = client.authMechList
      assertEquals(1, a.mechs.size)
      assertTrue(a.mechs.contains("CRAM-MD5"))
      assertEquals(1, server.getDecoder.getTransport.getNumberOfLocalConnections)
   }

   def testAuth(m: Method) {
      val props = new HashMap[String, String]
      val sc = Sasl.createSaslClient(Array("CRAM-MD5"), null, "hotrod", "localhost", props, new TestCallbackHandler("user", "realm", "password".toCharArray))
      val res = client.auth(sc)
      assertTrue(res.complete)
      assertEquals(1, server.getDecoder.getTransport.getNumberOfLocalConnections)
   }
   
   def testUnauthorizedOpCloseConnection(m: Method) {
      // Ensure the transport is clean
      val acceptedChannels : ChannelGroup = TestingUtil.extractField(server.getDecoder.getTransport, "acceptedChannels")
      acceptedChannels.close().awaitUninterruptibly()
      try {
        client.assertPutFail(m)
      } finally {
         assertEquals(0, server.getDecoder.getTransport.getNumberOfLocalConnections)
      }
   }

}

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/HotRodAuthenticationTest.java
package org.infinispan.server.hotrod;

import static org.assertj.core.api.Assertions.assertThat;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.startHotRodServer;
import static org.testng.AssertJUnit.assertEquals;
import static org.testng.AssertJUnit.assertTrue;

import java.lang.reflect.Method;
import java.util.HashMap;

import javax.security.sasl.Sasl;
import javax.security.sasl.SaslClient;
import javax.security.sasl.SaslException;

import org.infinispan.commons.util.SaslUtils;
import org.infinispan.manager.EmbeddedCacheManager;
import org.infinispan.server.core.security.simple.SimpleAuthenticator;
import org.infinispan.server.hotrod.configuration.HotRodServerConfigurationBuilder;
import org.infinispan.server.hotrod.test.HotRodTestingUtil;
import org.infinispan.server.hotrod.test.TestAuthMechListResponse;
import org.infinispan.server.hotrod.test.TestAuthResponse;
import org.infinispan.server.hotrod.test.TestCallbackHandler;
import org.infinispan.server.hotrod.test.TestResponse;
import org.infinispan.test.TestingUtil;
import org.testng.annotations.Test;

import io.netty.channel.group.ChannelGroup;

/**
 * Hot Rod server authentication test.
 *
 * @author Tristan Tarrant
 * @since 7.0
 */
@Test(groups = "functional", testName = "server.hotrod.HotRodAuthenticationTest")
public class HotRodAuthenticationTest extends HotRodSingleNodeTest {

   @Override
   public HotRodServer createStartHotRodServer(EmbeddedCacheManager cacheManager) {
      SimpleAuthenticator ssap = new SimpleAuthenticator();
      ssap.addUser("user", "realm", "password".toCharArray());
      HotRodServerConfigurationBuilder builder = new HotRodServerConfigurationBuilder();
      builder.authentication().enable()
            .sasl()
            .authenticator(ssap)
            .addAllowedMech("SCRAM-SHA-256")
            .serverName("localhost")
            .addMechProperty(Sasl.POLICY_NOANONYMOUS, "true");
      return startHotRodServer(cacheManager, HotRodTestingUtil.serverPort(), builder);
   }

   public void testAuthMechList() {
      TestAuthMechListResponse a = client().authMechList();
      assertEquals(1, a.mechs.size());
      assertTrue(a.mechs.contains("SCRAM-SHA-256"));
      assertEquals(1, server().getTransport().getNumberOfLocalConnections());
   }

   public void testAuth() throws SaslException {
      HashMap<String, String> props = new HashMap<>();
      SaslClient sc = SaslUtils.getSaslClientFactory(this.getClass().getClassLoader(), "SCRAM-SHA-256").createSaslClient(new String[]{"SCRAM-SHA-256"}, null, "hotrod", "localhost", props,
            new TestCallbackHandler("user", "realm", "password"));
      TestResponse res = client().auth(sc);
      assertThat(res).isInstanceOf(TestAuthResponse.class);
      assertEquals(1, server().getTransport().getNumberOfLocalConnections());
   }

   public void testUnauthorizedOpCloseConnection(Method m) {
      // Ensure the transport is clean
      ChannelGroup acceptedChannels =
            TestingUtil.extractField(server().getTransport(), "acceptedChannels");
      acceptedChannels.close().awaitUninterruptibly();
      try {
         client().assertPutFail(m);
      } finally {
         assertEquals(0, server().getTransport().getNumberOfLocalConnections());
      }
   }
}
